File: examples/poodle-surf/ios-objc/PoodleSurfObjC/Generic Views/GradientView.m
//
//  GradientView.m
//  PoodleSurfObjC
//
//  Created by Westin Newell on 5/10/19.
//  Copyright © 2019 Haiku. All rights reserved.
//

#import "GradientView.h"

NS_ASSUME_NONNULL_BEGIN

@interface GradientView ()

@property (nonatomic, readonly) CAGradientLayer *gradientLayer;

@end

@implementation GradientView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];

    _gradientLayer = [CAGradientLayer layer];
    _gradientLayer.frame = self.layer.bounds;

    [self.layer addSublayer:_gradientLayer];

    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];

    self.gradientLayer.frame = self.layer.bounds;
}

- (void)applyGradient:(DEZSimpleGradient *)gradient {
    self.gradientLayer.colors = @[
        (id)gradient.startColor.uiColor.CGColor,
        (id)gradient.endColor.uiColor.CGColor,
    ];
    self.gradientLayer.locations = @[@0, @1];
    self.gradientLayer.startPoint = (CGPoint){
        .x =  gradient.startPointX,
        .y = gradient.startPointX,
    };
    self.gradientLayer.endPoint = (CGPoint){
        .x = gradient.endPointX,
        .y = gradient.endPointX,
    };
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"initWithCoder: not implemented." userInfo:nil];
}

@end

NS_ASSUME_NONNULL_END

File: examples/poodle-surf/ios/PoodleSurf/Generic Views/GradientView.swift
//
//  Gradient.swift
//  PoodleSurf
//
//  Created by Westin Newell on 4/11/19.
//  Copyright © 2019 Haiku. All rights reserved.
//

import UIKit
import CoreGraphics

class GradientView: UIView {
    var gradient: Gradient? {
        didSet {
            guard let gradient = gradient else {
                gradientLayer.removeFromSuperlayer()
                return
            }

            gradientLayer.colors = gradient.colors.map { $0.color.cgColor }
            gradientLayer.locations = gradient.colors.map { NSNumber(floatLiteral: $0.location) }
            gradientLayer.startPoint = gradient.startPoint
            gradientLayer.endPoint = gradient.endPoint

            if gradientLayer.superlayer == nil {
                layer.addSublayer(gradientLayer)
            }
        }
    }

    override init(frame: CGRect) {
        super.init(frame: frame)

        gradientLayer.frame = layer.bounds
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        gradientLayer.frame = layer.bounds
    }

    private let gradientLayer = CAGradientLayer()

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) { fatalError("\(#function) not implemented.") }
}
